### ISSUE TITLE
Dockerfile support

### ISSUE THREAD

**Issue by proppy** [dokku/dokku#791]
In addition to buildpacks, it would be nice to support building the app from a Dockerfile if present (just like http://docs.deis.io/en/latest/using_deis/using-dockerfiles/#using-dockerfiles)

**Comment by techniq** [comment]
Looks like some progress on this front - https://github.com/progrium/dokku/pull/954

**Comment by josegonzalez** [comment]
As of version 0.3.15, Dokku should have full - and tested! - Dockerfile support.

**Comment by techniq** [comment]
@josegonzalez Great!

One thing I noticed in the example [`Dockerfile`](https://github.com/progrium/dokku/blob/master/tests/apps/dockerfile/Dockerfile) is that `ADD` is called before the `apt-get` statements, which is considered a bad practice since any changes to your app files will invalidate at that layer and thus cause those `RUN apt-get install ...` statements to have to be re-ran ever `git push dokku master` instead of leveraging the layer cache and thus speed up deployments greatly.

A better example might be:
```
FROM ubuntu:trusty
ENV LC_ALL C
ENV DEBIAN_FRONTEND noninteractive
ENV DEBCONF_NONINTERACTIVE_SEEN true
EXPOSE 5000

RUN apt-get install -y software-properties-common && add-apt-repository ppa:chris-lea/node.js && apt-get update
RUN apt-get install -y build-essential curl postgresql-client-9.3 nodejs git

COPY . /app
WORKDIR /app
RUN npm install

CMD npm start
```

**Comment by josegonzalez** [comment]
Feel free to PR that change to the test :)

**Comment by proppy** [comment]
Another good idea is to `ADD package.json` before `npm install` so dependencies get cached independently of the application files.

And doing `FROM node` would also simplify the `Dockerfile` further.

**Comment by techniq** [comment]
Done - https://github.com/progrium/dokku/pull/987

@proppy I've ran into issues trying to place the `ADD package.json` into a directory (say `/app`) and have it retrieved from the cache.  I've attempted to put it into `/tmp`, perform the `npm install`, and then move the `node_modules` directory under `/app` after adding the rest of the application files.  Might have something to do with using `WORKDIR`, but I've never been able to figure it out.   I agree on both those changes you suggest, I just wish I could get the `package.json` one to work.

**Comment by proppy** [comment]
@techniq feel free to take a look at how the official images are doing it (minus the `ONBUILD` bits):
https://github.com/joyent/docker-node/blob/master/0.12/onbuild/Dockerfile

**Comment by techniq** [comment]
@proppy 

Thanks, I just tried to duplicate that in my `Dockerfile`...
```
FROM ubuntu:trusty
ENV LC_ALL C
ENV DEBIAN_FRONTEND noninteractive
ENV DEBCONF_NONINTERACTIVE_SEEN true
EXPOSE 5000

RUN apt-get install -y software-properties-common && add-apt-repository ppa:chris-lea/node.js && apt-get update
RUN apt-get install -y build-essential curl postgresql-client-9.3 nodejs git

RUN mkdir -p /app
WORKDIR /app

COPY package.json /app/
RUN npm install
COPY . /app

CMD npm start
```
but the `COPY package.json /app/` step is not using / breaking the cache.

```
Step 0 : FROM ubuntu:trusty
 ---> 5ba9dab47459
Step 1 : ENV LC_ALL C
 ---> Using cache
 ---> 5f5c13a931d3
Step 2 : ENV DEBIAN_FRONTEND noninteractive
 ---> Using cache
 ---> 1ec730da88f0
Step 3 : ENV DEBCONF_NONINTERACTIVE_SEEN true
 ---> Using cache
 ---> 2650ef568dbb
Step 4 : EXPOSE 5000
 ---> Using cache
 ---> c07bb1b047fc
Step 5 : RUN apt-get install -y software-properties-common && add-apt-repository ppa:chris-lea/node.js && apt-get update
 ---> Using cache
 ---> d9f7e368b478
Step 6 : RUN apt-get install -y build-essential curl postgresql-client-9.3 nodejs git
 ---> Using cache
 ---> 82802567e015
Step 7 : RUN mkdir -p /app
 ---> Using cache
 ---> f80f79c1c96b
Step 8 : WORKDIR /app
 ---> Using cache
 ---> 61a31feba3a5
Step 9 : COPY package.json /app/
 ---> fed238a5f3ec
Removing intermediate container 138386a9938e
Step 10 : RUN npm install
 ---> Running in 9326452f937a
...
```

I don't want to blow up this issue troubleshooting this issue though (but it's always bothered me why I couldn't get it to work from what I've read)

**Comment by proppy** [comment]
@technik feel free to move the discussion to stack overflow and @mention me.

**Comment by techniq** [comment]
@proppy While working on creating a SO discussion on the `package.json` issue, I uncovered the reason for my issue (and it has some applicability with Dokku which is why I'm posting it here).  It turns out the cache is used as expected when I run `docker build .` repeated on a local directory, but if I git clone/copy/sync the application directory to another machine (say `git push dokku master` or use ansible to [sync](http://docs.ansible.com/synchronize_module.html) the folder over to another machine to build (ie. build an ARM image on a BeagleBone Black), the `COPY package.json /app/` line will always invalidate the cache.

I was able to consistently reproduce the issue without moving files between machines by making a copy of package.json (package-template.json) and then copying it back to package.json before creating a build (`cat package-template.json > package.json; docker build .`)

At this point I was going to log an issue with Docker, but then came across this existing issue - https://github.com/docker/docker/issues/3556 where it explains the changing of the modification time of the file is what is breaking the cache.  I found another open issue (https://github.com/docker/docker/issues/4351) that looks to be the one to watch for a possible solution in the future.

Lastly, I also came across an older [guide](http://www.clock.co.uk/blog/a-guide-on-how-to-cache-npm-install-with-docker) that mentioned this issue and suggested a bash script workaround to reset the mtime of package.json.  

_So the short of it_: `COPY package.json /app/` will invalidate the cache if `package.json`'s modified time changes between runs, such as from a git clone like Dokku.
